File: src/digraphx/neg_cycle_q.py
"""
Negative cycle detection for directed graphs.
1. Based on Howard's policy graph algorithm
2. Looking for more than one negative cycle
"""

from fractions import Fraction
from typing import (
    Callable,
    Dict,
    Generator,
    Generic,
    List,
    Mapping,
    MutableMapping,
    Tuple,
    TypeVar,
)

Node = TypeVar("Node")  # Hashable
Edge = TypeVar("Edge")  # Hashable
Domain = TypeVar("Domain", int, Fraction, float)  # Comparable Ring
Cycle = List[Edge]  # List of Edges


# The `NegCycleFinder` class implements Howard's method, a minimum cycle ratio algorithm, to find
# negative cycles in a directed graph.
class NegCycleFinder(Generic[Node, Edge, Domain]):
    """Negative Cycle Finder by Howard's method

    Howard's method is a minimum cycle ratio (MCR) algorithm that uses a policy
    iteration algorithm to find the minimum cycle ratio of a directed graph. The
    algorithm maintains a set of candidate cycles and iteratively updates the
    cycle with the minimum ratio until convergence. To detect negative cycles,
    Howard's method uses a cycle detection algorithm that is based on the
    Bellman-Ford relaxation algorithm. Specifically, the algorithm maintains a
    predecessor graph of the original graph and performs cycle detection on this
    graph using the Bellman-Ford relaxation algorithm. If a negative cycle is
    detected, the algorithm terminates and returns the cycle.
    """

    pred: Dict[Node, Tuple[Node, Edge]] = {}
    succ: Dict[Node, Tuple[Node, Edge]] = {}

    def __init__(self, gra: Mapping[Node, Mapping[Node, Edge]]) -> None:
        """
        The function initializes a graph object with an adjacency list.

        :param gra: The parameter `gra` is a mapping that represents an adjacency list. It is a
        dictionary-like object where the keys are nodes and the values are mappings of nodes to edges. Each
        edge represents a connection between two nodes in a directed graph

        :type gra: Mapping[Node, Mapping[Node, Edge]]
        """
        self.digraph = gra

    def find_cycle(self, point_to) -> Generator[Node, None, None]:
        """
        The `find_cycle` function is used to find a cycle in a policy graph and yields the start node of the
        cycle.

        :param point_to: The `point_to` parameter is a dictionary that represents the edges of a directed
        graph. Each key-value pair in the dictionary represents an edge from the key vertex to the value
        vertex

        Yields:
            Generator[Node, None, None]: a start node of the cycle

        Examples:
            >>> gra = {
            ...     "a0": {"a1": 7, "a2": 5},
            ...     "a1": {"a0": 0, "a2": 3},
            ...     "a2": {"a1": 1, "a0": 2},
            ... }
            >>> finder = NegCycleFinder(gra)
            >>> for cycle in finder.find_cycle(finder.pred):
            ...     print(cycle)
        """
        visited: Dict[Node, Node] = {}
        for vtx in filter(lambda vtx: vtx not in visited, self.digraph):
            utx = vtx
            while True:
                visited[utx] = vtx
                if utx not in point_to:
                    break
                utx, _ = point_to[utx]
                if utx in visited:
                    if visited[utx] == vtx:
                        yield utx
                    break

    def relax_pred(
        self,
        dist: MutableMapping[Node, Domain],
        get_weight: Callable[[Edge], Domain],
        update_ok: Callable[[MutableMapping[Node, Domain], Domain], bool],
    ) -> bool:
        """
        The `relax_pred` function updates the `dist` and `pred` dictionaries based on the current distances and
        weights of edges in a graph.

        :param dist: `dist` is a mutable mapping that represents the current distances from a source node to
        all other nodes in a graph. It is a mapping from nodes to their corresponding distances

        :type dist: MutableMapping[Node, Domain]

        :param get_weight: The `get_weight` parameter is a callable function that takes an `Edge` object as
        input and returns a value of type `Domain`. This function is used to calculate the weight or cost
        associated with an edge in the graph

        :type get_weight: Callable[[Edge], Domain]

        :param update_ok: The `update_ok` parameter is a function that determines whether an update to the
        distance `dist[vtx_v]` is allowed. It takes two arguments: the current value of `dist[vtx_v]` and
        the new value `d`. It should return `True` if the update is

        :return: a boolean value indicating whether any changes were made to the `dist` mapping and `pred` dictionary.
        """
        changed = False
        for utx, neighbors in self.digraph.items():
            for vtx, edge in neighbors.items():
                distance = dist[utx] + get_weight(edge)
                if dist[vtx] > distance and update_ok(dist[vtx], distance):
                    dist[vtx] = distance
                    self.pred[vtx] = (utx, edge)
                    changed = True
        return changed

    def relax_succ(
        self,
        dist: MutableMapping[Node, Domain],
        get_weight: Callable[[Edge], Domain],
        update_ok: Callable[[MutableMapping[Node, Domain], Domain], bool],
    ) -> bool:
        """
        The `relax_succ` function updates the `dist` and `succ` dictionaries based on the current distances and
        weights of edges in a graph.

        :param dist: `dist` is a mutable mapping that represents the current distances from a source node to
        all other nodes in a graph. It is a mapping from nodes to their corresponding distances

        :type dist: MutableMapping[Node, Domain]

        :param get_weight: The `get_weight` parameter is a callable function that takes an `Edge` object as
        input and returns a value of type `Domain`. This function is used to calculate the weight or cost
        associated with an edge in the graph

        :type get_weight: Callable[[Edge], Domain]

        :param update_ok: The `update_ok` parameter is a function that determines whether an update to the
        distance `dist[vtx_v]` is allowed. It takes two arguments: the current value of `dist[vtx_v]` and
        the new value `d`. It should return `True` if the update is

        :return: a boolean value indicating whether any changes were made to the `dist` mapping and `pred` dictionary.
        """
        changed = False
        for utx, neighbors in self.digraph.items():
            for vtx, edge in neighbors.items():
                distance = dist[vtx] - get_weight(edge)
                if dist[utx] < distance and update_ok(dist[utx], distance):
                    dist[utx] = distance
                    self.succ[utx] = (vtx, edge)
                    changed = True
        return changed

    def howard_pred(
        self,
        dist: MutableMapping[Node, Domain],
        get_weight: Callable[[Edge], Domain],
        update_ok: Callable[[MutableMapping[Node, Domain], Domain], bool],
    ) -> Generator[Cycle, None, None]:
        """
        The `howard_pred` function finds negative cycles in a graph and yields a list of cycles.

        :param dist: `dist` is a mutable mapping that maps each node in the graph to a domain value. The
        domain value represents the distance or cost from the source node to that particular node

        :type dist: MutableMapping[Node, Domain]

        :param get_weight: The `get_weight` parameter is a callable function that takes an `Edge` object as
        input and returns the weight of that edge

        :type get_weight: Callable[[Edge], Domain]

        :param update_ok: The `update_ok` parameter is a callable function that determines whether an update
        to the distance value of a vertex is allowed. It takes in three arguments: the current distance
        value of the vertex, the weight of the edge being considered for update, and the current distance
        value of the vertex at the other

        Examples:
            >>> gra = {
            ...     "a0": {"a1": 7, "a2": 5},
            ...     "a1": {"a0": 0, "a2": 3},
            ...     "a2": {"a1": 1, "a0": 2},
            ... }
            >>> dist = {vtx: 0 for vtx in gra}
            >>> def update_ok(dist, v) : return True
            >>> finder = NegCycleFinder(gra)
            >>> has_neg = False
            >>> for _ in finder.howard_pred(dist, lambda edge: edge, update_ok):
            ...     has_neg = True
            ...     break
            ...
            >>> has_neg
            False
        """
        self.pred = {}
        found = False
        while not found and self.relax_pred(dist, get_weight, update_ok):
            for vtx in self.find_cycle(self.pred):
                # Will zero cycle be found???
                assert self.is_negative(vtx, dist, get_weight)
                found = True
                yield self.cycle_list(vtx, self.pred)

    def howard_succ(
        self,
        dist: MutableMapping[Node, Domain],
        get_weight: Callable[[Edge], Domain],
        update_ok: Callable[[MutableMapping[Node, Domain], Domain], bool],
    ) -> Generator[Cycle, None, None]:
        """
        The `howard_succ` function finds negative cycles in a graph and yields a list of cycles.

        :param dist: `dist` is a mutable mapping that maps each node in the graph to a domain value. The
        domain value represents the distance or cost from the source node to that particular node

        :type dist: MutableMapping[Node, Domain]

        :param get_weight: The `get_weight` parameter is a callable function that takes an `Edge` object as
        input and returns the weight of that edge

        :type get_weight: Callable[[Edge], Domain]

        :param update_ok: The `update_ok` parameter is a callable function that determines whether an update
        to the distance value of a vertex is allowed. It takes in three arguments: the current distance
        value of the vertex, the weight of the edge being considered for update, and the current distance
        value of the vertex at the other

        Examples:
            >>> gra = {
            ...     "a0": {"a1": 7, "a2": 5},
            ...     "a1": {"a0": 0, "a2": 3},
            ...     "a2": {"a1": 1, "a0": 2},
            ... }
            >>> def update_ok(dist, v) : return True
            >>> dist = {vtx: 0 for vtx in gra}
            >>> finder = NegCycleFinder(gra)
            >>> has_neg = False
            >>> for _ in finder.howard_succ(dist, lambda edge: edge, update_ok):
            ...     has_neg = True
            ...     break
            ...
            >>> has_neg
            False
        """
        self.succ = {}
        found = False
        while not found and self.relax_succ(dist, get_weight, update_ok):
            for vtx in self.find_cycle(self.succ):
                # Will zero cycle be found???
                # assert self.is_negative(vtx, dist, get_weight)
                found = True
                yield self.cycle_list(vtx, self.succ)

    def cycle_list(self, handle: Node, point_to) -> Cycle:
        """
        The `cycle_list` function returns a list of edges that form a cycle in a graph, starting from a
        given node.

        :param handle: The `handle` parameter is a reference to a node in a graph. It represents the
        starting point of the cycle in the list

        :type handle: Node

        :param point_to: point_to is a dictionary that maps each graph node to the node it points to

        :return: a list of edges, which represents a cycle in a graph.
        """
        vtx = handle
        cycle = list()
        while True:
            utx, edge = point_to[vtx]
            cycle.append(edge)
            vtx = utx
            if vtx == handle:
                break
        return cycle

    def is_negative(
        self,
        handle: Node,
        dist: MutableMapping[Node, Domain],
        get_weight: Callable[[Edge], Domain],
    ) -> bool:
        """
        The `is_negative` function checks if a cycle list is negative by comparing the distances between
        nodes and the weights of the edges.

        :param handle: The `handle` parameter is a `Node` object that represents a vertex in a graph. It is
        used as a starting point to check for negative cycles in the graph

        :type handle: Node

        :param dist: `dist` is a mutable mapping that maps each node to its corresponding domain value. The
        domain value represents the distance from the starting node to the current node in a graph

        :type dist: MutableMapping[Node, Domain]

        :param get_weight: The `get_weight` parameter is a callable function that takes an `Edge` object as
        input and returns the weight of that edge

        :type get_weight: Callable[[Edge], Domain]

        :return: a boolean value.
        """
        vtx = handle
        # do while loop in C++
        while True:
            utx, edge = self.pred[vtx]
            if dist[vtx] > dist[utx] + get_weight(edge):
                return True
            vtx = utx
            if vtx == handle:
                break
        return False

File: rust_ai/src/neg_cycle_q.rs
//! Negative Cycle Finder with constraints (neg_cycle_q.rs)
//!
//! This module implements a Negative Cycle Finder for directed graphs using Howard's
//! method. The purpose of this code is to detect and find negative cycles in a
//! directed graph. A negative cycle is a cycle in the graph where the sum of the
//! edge weights is negative.

use std::collections::HashMap;
use std::marker::PhantomData;

use crate::types::{Cycle, Domain, Edge, Node};

/// Negative Cycle Finder with constraints using Howard's method
///
/// This struct implements Howard's method, which is a minimum cycle ratio (MCR) algorithm.
/// It works by maintaining a set of candidate cycles and iteratively updating them until
/// it finds the minimum cycle ratio or detects a negative cycle.
pub struct NegCycleFinderQ<N, E, D>
where
    N: Node,
    E: Edge,
    D: Domain,
{
    /// The directed graph where:
    /// - Keys are source nodes
    /// - Values are mappings of destination nodes to edges
    digraph: HashMap<N, HashMap<N, E>>,
    /// Dictionary to store predecessor information (node -> (predecessor_node, edge))
    pred: HashMap<N, (N, E)>,
    /// Dictionary to store successor information (node -> (successor_node, edge))
    succ: HashMap<N, (N, E)>,
    /// Marker for unused type parameter D
    _marker: PhantomData<D>,
}

impl<N, E, D> NegCycleFinderQ<N, E, D>
where
    N: Node,
    E: Edge,
    D: Domain,
{
    /// Initialize the negative cycle finder with a directed graph.
    ///
    /// # Arguments
    ///
    /// * `digraph` - A mapping representing a directed graph where:
    ///     - Keys are source nodes
    ///     - Values are mappings of destination nodes to edges
    pub fn new(digraph: HashMap<N, HashMap<N, E>>) -> Self {
        Self {
            digraph,
            pred: HashMap::new(),
            succ: HashMap::new(),
            _marker: PhantomData,
        }
    }

    /// Find cycles in the current predecessor graph using depth-first search.
    ///
    /// Uses a coloring algorithm (white/gray/black) to detect cycles.
    ///
    /// # Returns
    ///
    /// A vector of nodes that start cycles in the predecessor graph.
    pub fn find_cycle_pred(&self) -> Vec<N> {
        let mut visited: HashMap<N, N> = HashMap::new();
        let mut result = Vec::new();

        // Collect keys first to avoid borrowing issues
        let keys: Vec<N> = self.digraph.keys().cloned().collect();

        for vtx in keys {
            if visited.contains_key(&vtx) {
                continue;
            }

            let mut utx = vtx.clone();
            visited.insert(utx.clone(), vtx.clone());

            while let Some((pred_node, _)) = self.pred.get(&utx) {
                utx = pred_node.clone();

                if let Some(root) = visited.get(&utx) {
                    if root == &vtx {
                        result.push(utx.clone());
                    }
                    break;
                }

                visited.insert(utx.clone(), vtx.clone());
            }
        }

        result
    }

    /// Find cycles in the current successor graph using depth-first search.
    ///
    /// Uses a coloring algorithm (white/gray/black) to detect cycles.
    ///
    /// # Returns
    ///
    /// A vector of nodes that start cycles in the successor graph.
    pub fn find_cycle_succ(&self) -> Vec<N> {
        let mut visited: HashMap<N, N> = HashMap::new();
        let mut result = Vec::new();

        // Collect keys first to avoid borrowing issues
        let keys: Vec<N> = self.digraph.keys().cloned().collect();

        for vtx in keys {
            if visited.contains_key(&vtx) {
                continue;
            }

            let mut utx = vtx.clone();
            visited.insert(utx.clone(), vtx.clone());

            while let Some((succ_node, _)) = self.succ.get(&utx) {
                utx = succ_node.clone();

                if let Some(root) = visited.get(&utx) {
                    if root == &vtx {
                        result.push(utx.clone());
                    }
                    break;
                }

                visited.insert(utx.clone(), vtx.clone());
            }
        }

        result
    }

    /// Perform predecessor relaxation (relax_pred) operation.
    ///
    /// Updates distances based on predecessor edges.
    ///
    /// # Arguments
    ///
    /// * `dist` - Current distance estimates
    /// * `get_weight` - Function to get edge weights
    ///
    /// # Returns
    ///
    /// `true` if any distance was updated, `false` otherwise.
    pub fn relax_pred<F>(&mut self, dist: &mut HashMap<N, D>, get_weight: F) -> bool
    where
        F: Fn(&E) -> D,
    {
        let mut changed = false;

        for (utx, neighbors) in &self.digraph {
            let dist_u = dist.get(utx).cloned().unwrap_or_else(D::zero);

            for (vtx, edge) in neighbors {
                let weight = get_weight(edge);
                let distance = dist_u.clone() + weight;

                let dist_v = dist.entry(vtx.clone()).or_insert_with(D::zero);
                if *dist_v > distance {
                    *dist_v = distance;
                    self.pred.insert(vtx.clone(), (utx.clone(), edge.clone()));
                    changed = true;
                }
            }
        }

        changed
    }

    /// Perform successor relaxation (relax_succ) operation.
    ///
    /// Updates distances based on successor edges.
    ///
    /// # Arguments
    ///
    /// * `dist` - Current distance estimates
    /// * `get_weight` - Function to get edge weights
    ///
    /// # Returns
    ///
    /// `true` if any distance was updated, `false` otherwise.
    pub fn relax_succ<F>(&mut self, dist: &mut HashMap<N, D>, get_weight: F) -> bool
    where
        F: Fn(&E) -> D,
    {
        let mut changed = false;

        for (vtx, _predecessors) in &self.digraph {
            // We need to find edges that end at vtx
            for (utx, neighbors) in &self.digraph {
                if let Some(edge) = neighbors.get(vtx) {
                    let dist_u = dist.get(utx).cloned().unwrap_or_else(D::zero);
                    let weight = get_weight(edge);
                    let distance = dist_u.clone() + weight;

                    let dist_v = dist.entry(vtx.clone()).or_insert_with(D::zero);
                    if *dist_v > distance {
                        *dist_v = distance;
                        self.succ.insert(utx.clone(), (vtx.clone(), edge.clone()));
                        changed = true;
                    }
                }
            }
        }

        changed
    }

    /// Reconstruct the cycle starting from the given node using predecessor links.
    ///
    /// # Arguments
    ///
    /// * `handle` - The starting node of the cycle
    /// * `pred_map` - Predecessor mapping to use
    ///
    /// # Returns
    ///
    /// List of edges forming the cycle in order.
    pub fn cycle_list_pred(&self, handle: &N, pred_map: &HashMap<N, (N, E)>) -> Cycle<E> {
        let mut cycle = Vec::new();
        let mut vtx = handle.clone();

        loop {
            let (utx, edge) = pred_map.get(&vtx).expect("Node not in predecessor graph");
            cycle.push(edge.clone());
            vtx = utx.clone();

            if &vtx == handle {
                break;
            }
        }

        cycle
    }

    /// Reconstruct the cycle starting from the given node using successor links.
    ///
    /// # Arguments
    ///
    /// * `handle` - The starting node of the cycle
    /// * `succ_map` - Successor mapping to use
    ///
    /// # Returns
    ///
    /// List of edges forming the cycle in order.
    pub fn cycle_list_succ(&self, handle: &N, succ_map: &HashMap<N, (N, E)>) -> Cycle<E> {
        let mut cycle = Vec::new();
        let mut vtx = handle.clone();

        loop {
            let (next_vtx, edge) = succ_map.get(&vtx).expect("Node not in successor graph");
            cycle.push(edge.clone());
            vtx = next_vtx.clone();

            if &vtx == handle {
                break;
            }
        }

        cycle
    }

    /// Check if the cycle starting at 'handle' is negative using predecessor links.
    ///
    /// # Arguments
    ///
    /// * `handle` - Starting node of the cycle to check
    /// * `dist` - Current distance estimates
    /// * `get_weight` - Function to get edge weights
    /// * `pred_map` - Predecessor mapping to use
    ///
    /// # Returns
    ///
    /// `true` if the cycle is negative, `false` otherwise.
    pub fn is_negative_pred<F>(
        &self,
        handle: &N,
        dist: &HashMap<N, D>,
        get_weight: F,
        pred_map: &HashMap<N, (N, E)>,
    ) -> bool
    where
        F: Fn(&E) -> D,
    {
        let mut vtx = handle.clone();

        loop {
            let (utx, edge) = pred_map.get(&vtx).expect("Node not in predecessor graph");
            let weight = get_weight(edge);

            let dist_v = dist.get(&vtx).cloned().unwrap_or_else(D::zero);
            let dist_u = dist.get(utx).cloned().unwrap_or_else(D::zero);

            if dist_v > dist_u.clone() + weight {
                return true;
            }

            vtx = utx.clone();
            if &vtx == handle {
                break;
            }
        }

        false
    }

    /// Check if the cycle starting at 'handle' is negative using successor links.
    ///
    /// # Arguments
    ///
    /// * `handle` - Starting node of the cycle to check
    /// * `dist` - Current distance estimates
    /// * `get_weight` - Function to get edge weights
    /// * `succ_map` - Successor mapping to use
    ///
    /// # Returns
    ///
    /// `true` if the cycle is negative, `false` otherwise.
    pub fn is_negative_succ<F>(
        &self,
        handle: &N,
        dist: &HashMap<N, D>,
        get_weight: F,
        succ_map: &HashMap<N, (N, E)>,
    ) -> bool
    where
        F: Fn(&E) -> D,
    {
        let mut vtx = handle.clone();

        loop {
            let (next_vtx, edge) = succ_map.get(&vtx).expect("Node not in successor graph");
            let weight = get_weight(edge);

            let dist_v = dist.get(&vtx).cloned().unwrap_or_else(D::zero);
            let dist_u = dist.get(next_vtx).cloned().unwrap_or_else(D::zero);

            if dist_v > dist_u.clone() + weight {
                return true;
            }

            vtx = next_vtx.clone();
            if &vtx == handle {
                break;
            }
        }

        false
    }

    /// Main algorithm to find negative cycles using Howard's method with predecessor relaxation.
    ///
    /// # Arguments
    ///
    /// * `dist` - Initial distance estimates
    /// * `get_weight` - Function to get edge weights
    ///
    /// # Returns
    ///
    /// A vector of found negative cycles, each as a list of edges.
    pub fn howard_pred<F>(&mut self, dist: &mut HashMap<N, D>, get_weight: F) -> Vec<Cycle<E>>
    where
        F: Fn(&E) -> D + Clone,
    {
        let mut cycles = Vec::new();
        self.pred.clear();
        let mut found = false;

        while !found && self.relax_pred(dist, &get_weight) {
            for vtx in self.find_cycle_pred() {
                assert!(self.is_negative_pred(&vtx, dist, &get_weight, &self.pred));
                found = true;
                cycles.push(self.cycle_list_pred(&vtx, &self.pred));
            }
        }

        cycles
    }

    /// Main algorithm to find negative cycles using Howard's method with successor relaxation.
    ///
    /// # Arguments
    ///
    /// * `dist` - Initial distance estimates
    /// * `get_weight` - Function to get edge weights
    ///
    /// # Returns
    ///
    /// A vector of found negative cycles, each as a list of edges.
    pub fn howard_succ<F>(&mut self, dist: &mut HashMap<N, D>, get_weight: F) -> Vec<Cycle<E>>
    where
        F: Fn(&E) -> D + Clone,
    {
        let mut cycles = Vec::new();
        self.succ.clear();
        let mut found = false;

        while !found && self.relax_succ(dist, &get_weight) {
            for vtx in self.find_cycle_succ() {
                assert!(self.is_negative_succ(&vtx, dist, &get_weight, &self.succ));
                found = true;
                cycles.push(self.cycle_list_succ(&vtx, &self.succ));
            }
        }

        cycles
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_neg_cycle_finder_q_new() {
        let digraph: HashMap<&str, HashMap<&str, i32>> = HashMap::new();
        let finder: NegCycleFinderQ<&str, i32, i32> = NegCycleFinderQ::new(digraph);
        assert!(finder.pred.is_empty());
        assert!(finder.succ.is_empty());
    }

    #[test]
    fn test_relax_pred() {
        let mut digraph = HashMap::new();
        let mut neighbors = HashMap::new();
        neighbors.insert("b", 1);
        neighbors.insert("c", 4);
        digraph.insert("a", neighbors);

        let mut neighbors = HashMap::new();
        neighbors.insert("c", 2);
        digraph.insert("b", neighbors);

        let mut neighbors = HashMap::new();
        neighbors.insert("a", -5);
        digraph.insert("c", neighbors);

        let mut finder: NegCycleFinderQ<&str, i32, i32> = NegCycleFinderQ::new(digraph);
        let mut dist = HashMap::new();
        dist.insert("a", 0);
        dist.insert("b", 1000); // Use a large but safe initial value
        dist.insert("c", 1000);

        let changed = finder.relax_pred(&mut dist, |edge| *edge);
        assert!(changed);
        assert_eq!(dist["b"], 1);
        assert_eq!(dist["c"], 3);
    }
}
